File: day19/series_of_tubes.py
#!/usr/bin/env python3
from sys import argv
import re

# parse the maze
maze = []
filename = argv[1]
with open(filename) as f:
    for line in f:
        line = line.rstrip("\n")
        maze.append(line)

# find the opening position
row = 0
col = 0
for i in range(len(maze[0])):
    if maze[0][i] == '|':
        col = i
        break

# walk through the maze
drow, dcol = 1,0 # down
done = False
letters = ""
steps = 0
while not done:
    row += drow
    col += dcol
    steps += 1
    val = maze[row][col]
    if val == ' ': # ran out of space, so finished!
        done = True
    elif re.match('[A-Z]', val):
        letters += val
    elif val == '+': # turn
        if dcol == 0: # look left and right
            if maze[row][col-1] != ' ':
                drow, dcol = 0, -1 # left
            else:
                drow, dcol = 0, 1  # right
        else: # look up and down
            if maze[row-1][col] != ' ':
                drow, dcol = -1, 0 # up
            else:
                drow, dcol = 1, 0  # down

print("result1:", letters)
print("result2:", steps)

File: day19/series_of_tubes.pl
#!/usr/bin/env perl
use strict;
use warnings;
use feature qw(:5.24);
use experimental qw(signatures);

# read in the maze
my @maze;
while (<>) {
    chomp;
    push @maze, [split //];
}

# find the opening position
my $len = @{$maze[0]};
my $row = 0;
my $col;
for my $i (0..$len-1) {
    if ($maze[0][$i] eq '|') {
        $col = $i;
        last;
    }
}

# walk through the maze
my ($drow, $dcol) = (1, 0); # down
my $done = 0;
my $letters = "";
my $steps = 0;
while (!$done) {
    $row += $drow;
    $col += $dcol;
    $steps++;
    my $val = $maze[$row][$col];
    if ($val eq ' ') { # ran out of space, so finished!
        $done = 1;
    } elsif ($val =~ /[A-Z]/) {
        $letters .= $val;
    } elsif ($val eq '+') { # turn!
        if ($dcol == 0) { # look left and right
            if ($maze[$row][$col-1] ne ' ') {
                ($drow, $dcol) = (0, -1); # left
            } else {
                ($drow, $dcol) = (0, 1);  # right
            }
        } else { # look up and down
            if ($maze[$row-1][$col] ne ' ') {
                ($drow, $dcol) = (-1, 0); # up
            } else {
                ($drow, $dcol) = (1, 0);  #down
            }
        }
    }
}

say "result1: $letters";
say "result2: $steps";
